### DISCUSSION TITLE
How to access user request data on Insert with Supabase default values

### DISCUSSION THREAD

**Discussion by bennik88** [supabase/supabase#9448] [Questions]
## Problem
When I'm trying to access the user request data in a `FUNCTION` that is being called after a `BEFORE` trigger the data will always be in the `NEW` variable. This means that the Supabase default value will always overwrite the user request data before I can access it from within the `NEW` variable. The `OLD` value will always be `NULL` on inserts. I'm already using a `BEFORE` trigger to ensure to catch it as early as possible.

## Question
As Supabase is overwriting the `NEW` value how can I check what the user has sent in his request on insert?

## Example
If the user tries to add data to the request of the insert statement he is not supposed to supabase default values will overwrite the default values into the `NEW` attribute.
On insert, the `OLD` attribute will always be `NULL`.

The function will always raise an exception as the `NEW` variable is never `NULL` due to the supabase default values.
```sql
CREATE OR REPLACE FUNCTION validate_to_do()
RETURNS TRIGGER AS
$$
    BEGIN
    -- Security check that no id, created_at, or updated_at information is passed in
    IF NEW.id IS NOT NULL
    OR NEW.created_at IS NOT NULL
    OR NEW.last_updated_at IS NOT NULL
    THEN
        -- Log all values as exception to make ensure the describes behaviour
        RAISE EXCEPTION 'Cannot set id, created_at or last_updated_at. NEW! id: %, created_at: %, last_updated_at: %. OLD! id: %, created_at: %, last_updated_at: %', NEW.id, NEW.created_at, NEW.last_updated_at, OLD.id, OLD.created_at, OLD.last_updated_at;
        END IF;
        RETURN NEW;
    END;
$$
LANGUAGE plpgsql;

CREATE TRIGGER on_create_to_do
    BEFORE INSERT ON public.to_do
    FOR EACH ROW
        EXECUTE PROCEDURE public.validate_to_do();
```

**Comment by GaryAustin1** [comment]
The new value is only overwritten by default if the column is not present in the insert.  Otherwise you get the user value.  Default only applies if the column is not supplied.
So 
insert([{othercol:'cat'}]) new.col would equal the default value
insert([{col:null,othercol:'cat'}]) new.col = null
insert([{col:'dog',othercol:'cat'}]) new.col = 'dog'

You can use `raise log 'new value = %', new.col;`  to see the values in the Postgres log on functions.

**Reply by bennik88** [reply]
That’s absolutely right and matches to my tests.
The issue now as stated above is how would I now find out if the `NEW` value came from an user or from the set supabase default value? 
At the end for example, to ensure the user can’t overwrite the default value. In a best case without overwriting it with another default value.

**Reply by GaryAustin1** [reply]
Yes, you just need to force the default value again.  Assigning null won't work as that will override the default and although I have not double checked the docs, I don't know how you would get rid of new.col in the new object which is the only way to get it to use default otherwise, and it may already be too late in the process for that to work as default was already assigned.

**Reply by bennik88** [reply]
So that essentially means when using default values there is no way to check whether the information came from a user or the default value, right?

I guess then removing the default value completely and setting it in a function after a trigger is a way to go. As otherwise the default value then is just wasting performance on every insert or update when I overwrite it anyway.
Or am I missing anything and it would be a best practice to set it two times once in a table and once in a trigger function?

**Reply by GaryAustin1** [reply]
I believe that declaring a default value is a waste for what you are doing with a trigger on it.  

I honestly though can't comment on best practices.

You might google something like `postgres set default insert trigger` and see if there is more info.

**Reply by GaryAustin1** [reply]
Related:
https://stackoverflow.com/questions/7797425/how-to-force-the-default-value-of-a-column-in-postgresql

**Reply by bennik88** [reply]
To summarise your stack overflow link:
- Triggers are the way to go in such a case.
- Another good practice seems to raise an error on invalid data. This is what I wanted in the first place but since I can't compare the default value from supabase with the user data, as the default value will be already overwritten when I try to access it, comparing the values won't work. After triggers wouldn't work eigther as I work with timestamps and these would likely have changed in the meantime. Instead revoking it would be the second-best bet from what I can see. I've not done this or tested it yet so I can't speak about bottlenecks or issues. But using `revoke insert (id)...` at least also sounds like a good option if you prefer throwing an error if somebody tries to do suspicious things.
